lib/core/test/unit/Cardano/Wallet/Primitive/ModelSpec.hs

```diff
@@ -2387,3 +2440,388 @@ outputsCreatedByTx tx
+--    global_utxo_1 -> block_1_2 ->
+--    ...
+--    global_utxo_i -> block_i_j ->
+--    global_utxo_j
+-- @
+--
+-- In turn, each block contains a series of transactions that each affect the
+-- global UTxO set. The effect of applying a single block to the global UTxO
+-- set is equivalent to applying all of the contained transactions in order
+-- to the global UTxO set.
+--
+-- In general, only a subset of transactions within a block sequence will be
+-- relevant to a particular wallet, and only a subset of entries within each
+-- successive iteration of the global UTxO set will be relevant to that wallet.
+--
+-- Values of 'BlockSeq' are more convenient to generate and shrink than values
+-- of 'BlockData', but can easily be transformed into values of 'BlockData' in
+-- order to test properties of the 'applyBlocks' function.
+--
+-- Usage:
+--
+--  1. Use function 'genBlockSeq' to generate a 'BlockSeq' value.
+--  2. Use function 'blockSeqToBlockData' to convert a 'BlockSeq' value to a
+--     value of 'BlockData' suitable for use with the 'applyBlocks' function.
+--  3. Use function 'shrinkBlockSeq' to shrink to a minimal counterexample.
+--
+data BlockSeq = BlockSeq
+    { initialBlockHeight
+        :: Quantity "block" Word32
+    , initialSlotNo
+        :: SlotNo
+    , shrinkableTxSeq
+        :: ShrinkableTxSeq
+    }
+    deriving (Eq, Generic, Show)
+
+instance Arbitrary BlockSeq where
+    arbitrary = genBlockSeq
+    shrink = shrinkBlockSeq
+
+genBlockSeq :: Gen BlockSeq
+genBlockSeq = BlockSeq
+    <$> fmap toEnum (choose (0, 100))
+    <*> fmap toEnum (choose (0, 100))
+    <*> genTxSeq genUTxO genAddress
+
+shrinkBlockSeq :: BlockSeq -> [BlockSeq]
+shrinkBlockSeq = genericRoundRobinShrink
+    <@> shrinkMap toEnum fromEnum
+    <:> shrinkMap toEnum fromEnum
+    <:> shrinkTxSeq
+    <:> Nil
+
+-- | Converts a 'BlockSeq' to a value of 'BlockData' that is suitable for use
+--   with the 'applyBlocks' function.
+--
+-- This function fills every slot with exactly one block, leaving no gaps.
+--
+blockSeqToBlockData :: BlockSeq -> BlockData m addr tx state
+blockSeqToBlockData = List . blockSeqToBlockList
+  where
+    blockSeqToBlockList :: BlockSeq -> NonEmpty Block
+    blockSeqToBlockList blockSeq =
+        NE.fromList $ getZipList $ makeBlock
+            <$> ZipList (enumFrom $ blockSeq & initialBlockHeight)
+            <*> ZipList (enumFrom $ blockSeq & initialSlotNo)
+            <*> ZipList (NE.toList $ TxSeq.toTxGroupList txSeq)
+      where
+        txSeq :: TxSeq
+        txSeq = blockSeqToTxSeq blockSeq
+
+        -- Makes a block using dummy values for fields that are not relevant
+        -- to our expectations of 'applyBlocks'.
+        --
+        makeBlock :: Quantity "block" Word32 -> SlotNo -> [Tx] -> Block
+        makeBlock blockHeight slotNo transactions = Block
+            { header = BlockHeader
+                { blockHeight
+                , slotNo
+                , headerHash = Hash ""
```

Comment:
I suggest to consider using something like `error "headerHash is irrelevant"` to fail fast and precise if assumption doesn't hold.

Reply:
Actually, you already use it below, `shouldNotEvaluate "Block.headerHash"`

Reply:
@Unisay I did do this originally!

However, then I discovered that `headerHash` is a strict field of `BlockHeader`, and `applyBlocks` pattern matches on values of `BlockHeader` in order to construct values of `ChainPoint`.

So using `error` here is not an option, unfortunately, unless we change `applyBlocks` itself.